swap/swap.go

```diff
@@ -1,98 +1,636 @@
+	contract "github.com/ethersphere/swarm/contracts/swap"
 	"github.com/ethersphere/swarm/log"
 	"github.com/ethersphere/swarm/p2p/protocols"
 	"github.com/ethersphere/swarm/state"
 )
 
-// SwAP Swarm Accounting Protocol
+// ErrInvalidChequeSignature indicates the signature on the cheque was invalid
+var ErrInvalidChequeSignature = errors.New("invalid cheque signature")
+
+// Swap represents the Swarm Accounting Protocol
 // a peer to peer micropayment system
 // A node maintains an individual balance with every peer
 // Only messages which have a price will be accounted for
 type Swap struct {
-	stateStore state.Store        //stateStore is needed in order to keep balances across sessions
-	lock       sync.RWMutex       //lock the balances
-	balances   map[enode.ID]int64 //map of balances for each peer
+	api                 PublicAPI
+	store               state.Store          // store is needed in order to keep balances and cheques across sessions
+	accountingLock      sync.RWMutex         // lock for data consistency in accounting-related functions
+	balances            map[enode.ID]int64   // map of balances for each peer
+	balancesLock        sync.RWMutex         // lock for balances map
+	cheques             map[enode.ID]*Cheque // map of cheques for each peer
+	chequesLock         sync.RWMutex         // lock for cheques map
+	peers               map[enode.ID]*Peer   // map of all swap Peers
+	peersLock           sync.RWMutex         // lock for peers map
+	backend             contract.Backend     // the backend (blockchain) used
+	owner               *Owner               // contract access
+	params              *Params              // economic and operational parameters
+	contract            swap.Contract        // reference to the smart contract
+	oracle              PriceOracle          // the oracle providing the ether price for honey
+	paymentThreshold    int64                // balance difference required for sending cheque
+	disconnectThreshold int64                // balance difference required for dropping peer
+}
+
+// Owner encapsulates information related to accessing the contract
+type Owner struct {
+	Contract   common.Address    // address of swap contract
+	address    common.Address    // owner address
+	privateKey *ecdsa.PrivateKey // private key
+	publicKey  *ecdsa.PublicKey  // public key
+}
+
+// Params encapsulates param
+type Params struct {
+	InitialDepositAmount uint64
+}
+
+// NewParams returns a Params struct filled with default values
+func NewParams() *Params {
+	return &Params{
+		InitialDepositAmount: DefaultInitialDepositAmount,
+	}
 }
 
 // New - swap constructor
-func New(stateStore state.Store) (swap *Swap) {
-	swap = &Swap{
-		stateStore: stateStore,
-		balances:   make(map[enode.ID]int64),
+func New(stateStore state.Store, prvkey *ecdsa.PrivateKey, contract common.Address, backend contract.Backend) *Swap {
+	return &Swap{
+		store:               stateStore,
+		balances:            make(map[enode.ID]int64),
```

Comment:
these are not needed if  accessors are defined on the Peer

Reply:
should be solved under #1604.

Reply:
relevant issue: #1604.